design-proposals/sriov-interfaces-mapping.md

```diff
@@ -0,0 +1,714 @@
+> The vfio-pci driver is an userspace driver, when a VF is bound to it, the VF is no longer recognized by the kernel and therefore will not be presented by iplink (‘ip a’ command).
+
+# Proposal
+## General Flow
+The proposal is based on information multus-cni provides as an annotation on the virt-launcher pod. The objective is to pass down the relevant information to the virt-launcher application, so it in turn can consume it and create the correct domain configuration.
+
+In order to pass the information from the pod to the virt-launcher application, the following step are to be taken:
+Create a generic annotation for passing information from the pod manifest to the application that runs in it. This annotation is using the downward api to make it available to the application.
+Given the networks-status annotation, process the content and compose structure that contains a 1:1 mapping between a (VMI) network from it and the device PCI address. This data is added to the generic annotation from earlier.
+Once the virt-launcher application requires this network-to-pci information, it can read the content.
+In the context of SR-IOV, virt-launcher will use this data when creating the domain for the first time and when hotplug-ing the devices.
+
+The following sections go into more details.
+
+## Utilizing “k8s.v1.cni.cncf.io/networks-status” annotation
+This annotation includes the information required to do one-to-one mapping between a VM interface and VF PCI address:
+
+### Example
+```yaml
+k8s.v1.cni.cncf.io/networks-status: |-
+    [
+        {
+            "name": "kindnet",
+            "interface": "eth0",
+            "ips": [
+                "10.244.2.131"
+            ],
+            "mac": "82:cf:7c:98:43:7e",
+            "default": true,
+            "dns": {}
+        },{
+            "name": "default/sriov-network-vlan100",
+            "interface": "net1",
+            "dns": {},
+            "device-info": {
+                "type": "pci",
+                "version": "1.0.0",
+                "pci": {
+                    "pci-address": "0000:04:02.5"
+                }
+            }
+        },{
+            "name": "default/sriov-network-vlan200",
+            "interface": "net2",
+            "dns": {},
+            "device-info": {
+                "type": "pci",
+                "version": "1.0.0",
+                "pci": {
+                    "pci-address": "0000:04:02.2"
+                }
+            }
+        }
+    ]
+```
+
+> **_Note_**: Multus provides PCI device information since version v3.7.
+
+> **_Note_**: in order to maintain backward compatibility, when the annotation is missing, virt-launcher will fall back to naive mapping and use the legacy environment variable.
+
+> **_Note_**: Once a Pod reaches Running state, the CNI operation is finished successfully and the annotation will present on the Pod.
+
+### Annotations Exposure Method
```

Comment:
In general I would stay away from annotations whenever possible. Multus has to make use of them, but we should not use them if not absolutely necessary. Neither on the VMI, nor on the Pod.

Reply:
In general totaly agree, but in this case inevitable because  there is no way to get the network-to-pci mapping from an API other than from Multus network-status annotation.

Note that we use `k8s.v1.cni.cncf.io/networks` annotation in order to connect the pod to secondary networks.
The `k8s.v1.cni.cncf.io/networks-status` annotation is the result of that operation.

Reply:
Yes, if we have to integrate with third-party apps which use it there may be no other way. I am not against using existing multus annotations. I am just arguing about adding additional ones where we already have better communication pathes.